### PR TITLE
Optimize local map read

### PR DESCRIPTION
**Pull request by yamanogluberk**
Optimizing `ReadMapOrCacheP` for Local Map Reading

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 1]
```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
...
+
+            CompletableFuture<MapEntriesWithCursor> f = new CompletableFuture<>();
+            read0(f, partitionId, pointers);
+            return f;
+        }
+
+        private void read0(CompletableFuture<MapEntriesWithCursor> f, int partitionId, IterationPointer[] pointers) {
+            try {
+                boolean isOwned = mapServiceContext.getOrInitCachedMemberPartitions().contains(partitionId);
+                int migrationStamp = mapServiceContext.getService().getMigrationStamp();
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:518]: This should go after `if (isOwned)`, it's not applicable to reading using an operation

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 2]
```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
...
+                if (isOwned) {
+                    if (!validateMigrationStamp(migrationStamp)) {
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:520]: We don't need to validate right after getting the stamp, but after the reading.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 3]
```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
     }
 
     static class LocalMapReader
-            extends Reader<InternalCompletableFuture<MapEntriesWithCursor>, MapEntriesWithCursor, Entry<Data, Data>> {
+            extends Reader<CompletableFuture<MapEntriesWithCursor>, MapEntriesWithCursor, Entry<Data, Data>> {
 
         private final MapProxyImpl mapProxyImpl;
+        private final MapServiceContext mapServiceContext;
+        private final NodeEngineImpl nodeEngine;
+        private final int retryDelay = 100;
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:488]:
```suggestion
        private static final int RETRY_DELAY = 100;
```

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 4]
```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
...
+            this.mapProxyImpl = (MapProxyImpl) hzInstance.getMap(mapName);
+            nodeEngine = getNodeEngine(hzInstance);
+            MapService service = nodeEngine.getService(MapService.SERVICE_NAME);
+            mapServiceContext = service.getMapServiceContext();
         }
 
         @Nonnull @Override
-        public InternalCompletableFuture<MapEntriesWithCursor> readBatch(int partitionId, IterationPointer[] pointers) {
+        public CompletableFuture<MapEntriesWithCursor> readBatch(int partitionId, IterationPointer[] pointers) {
+            boolean isHD = mapProxyImpl.getMapConfig().getInMemoryFormat().equals(InMemoryFormat.NATIVE);
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:506]: We can read this config once and store in a field

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 5]
```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
...
+                int migrationStamp = mapServiceContext.getService().getMigrationStamp();
+                if (isOwned) {
+                    if (!validateMigrationStamp(migrationStamp)) {
+                        scheduleForLater(f, partitionId, pointers);
+                        return;
+                    }
+
+                    MapEntriesWithCursor result = accessRecordStore(partitionId, pointers);
+
+                    if (!validateMigrationStamp(migrationStamp)) {
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:527]: If you invert the if-else clause, it's a bit easier to read.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java [hunk 6]
```diff
@@ -90,6 +90,23 @@ public void stressTest_hash() throws InterruptedException {
         assertThat(mutatorException.get()).isNull();
     }
 
+    @Test(timeout = 600_000)
+    public void stressTest_fullFetch() throws InterruptedException {
+        List<Row> expected = new ArrayList<>();
+        for (int i = 0; i <= ITEM_COUNT / 5; i++) {
+            map.put(i, 1);
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java:97]: We should use `IMap.putAll` for such big intializations, it's much faster.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java [hunk 7]
```diff
@@ -90,6 +90,23 @@ public void stressTest_hash() throws InterruptedException {
         assertThat(mutatorException.get()).isNull();
     }
 
+    @Test(timeout = 600_000)
+    public void stressTest_fullFetch() throws InterruptedException {
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java:94]: I'd name the three tests in this class this way and would add them in this order:
- `stressTest_noIndex`
- `stressTest_hashIndex`
- `stressTest_sortedIndex`

**Review by viliam-durina — COMMENTED** [review]
Looks good overall

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 8]
```diff
@@ -44,7 +44,6 @@
     @Test
     public void testColumn() {
         ExpressionType<?>[] allTypes = ExpressionTypes.all();
-
```
Comment (hakanmemisoglu) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:47]: Not important but you can remove this change.

**Review by hakanmemisoglu — APPROVED** [review]
Looks good to me after @viliam-durina's suggestions.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java [hunk 9]
```diff
@@ -472,27 +480,95 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
...
+            } catch (Throwable t) {
+                f.completeExceptionally(t);
+            }
+        }
+
+        private void scheduleForLater(CompletableFuture<MapEntriesWithCursor> f, int partitionId, IterationPointer[] pointers) {
+            nodeEngine.getExecutionService().schedule(() -> read0(f, partitionId, pointers), RETRY_DELAY, TimeUnit.MILLISECONDS);
+        }
+
+        private MapEntriesWithCursor accessRecordStore(int partitionId, IterationPointer[] pointers) {
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java:550]: Better name might be `readFromRecordStore`

**Review by viliam-durina — APPROVED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java [hunk 10]
```diff
@@ -91,12 +114,14 @@ public void stressTest_hash() throws InterruptedException {
...
-    public void stressTest_sorted() throws InterruptedException {
+    public void stressTest_sortedIndex() throws InterruptedException {
         List<Row> expected = new ArrayList<>();
+        Map<Integer, Integer> temp = new HashMap<>();
         for (int i = 0; i <= ITEM_COUNT; i++) {
-            map.put(i, i);
+            temp.put(i, i);
             expected.add(new Row(ITEM_COUNT - i, ITEM_COUNT - i));
         }
+        map.putAll(temp);
```
[the remark this answers is not in the document]
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java:124]: The expected rows are different from each other and we would need that for loop anyways if I were to extract the map filling logic. Do you think I should extract it?

[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql_slow/JetMapScanMigrationStressTest.java:124]: If it's not exactly the same I leave it up to you - whatever you see as more readable/maintainable...

**Comment by yamanogluberk** [comment]
run-lab-run

**Comment by yamanogluberk** [comment]
run-lab-run

**Comment by yamanogluberk** [comment]
run-lab-run

**Pull request merged**